onnxruntime/contrib_ops/cpu/murmur_hash3.cc

```diff
@@ -0,0 +1,203 @@
+// Finalization mix - force all bits of a hash block to avalanche
+
+FORCE_INLINE uint32_t fmix(uint32_t h) {
+  h ^= h >> 16;
+  h *= 0x85ebca6b;
+  h ^= h >> 13;
+  h *= 0xc2b2ae35;
+  h ^= h >> 16;
+
+  return h;
+}
+
+//----------
+
+FORCE_INLINE uint64_t fmix(uint64_t k) {
+  k ^= k >> 33;
+  k *= BIG_CONSTANT(0xff51afd7ed558ccd);
+  k ^= k >> 33;
+  k *= BIG_CONSTANT(0xc4ceb9fe1a85ec53);
+  k ^= k >> 33;
+
+  return k;
+}
+
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    MurmurHash3,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder()
+        .TypeConstraint("T1", std::vector<MLDataType>{DataTypeImpl::GetTensorType<int32_t>(),
+                                                      DataTypeImpl::GetTensorType<uint32_t>(),
+                                                      DataTypeImpl::GetTensorType<std::string>()})
+        .TypeConstraint("T2", std::vector<MLDataType>{DataTypeImpl::GetTensorType<int32_t>(),
+                                                      DataTypeImpl::GetTensorType<uint32_t>()}),
+    MurmurHash3);
+
+void MurmurHash3::MurmurHash3_x86_32(const void* key, int len, uint32_t seed, void* out) const {
+
+  const uint8_t* data = (const uint8_t*)key;
+  const int nblocks = len / 4;
+  uint32_t h1 = seed;
+  uint32_t c1 = 0xcc9e2d51;
+  uint32_t c2 = 0x1b873593;
+
+  //----------
+  // body
+  const uint32_t* blocks = (const uint32_t*)(data + (int64_t)nblocks * 4);
+
+  for (int i = -nblocks; i; i++) {
+    uint32_t k1 = getblock(blocks, i);
+
+    k1 *= c1;
+    k1 = ROTL32(k1, 15);
+    k1 *= c2;
+
+    h1 ^= k1;
+    h1 = ROTL32(h1, 13);
+    h1 = h1 * 5 + 0xe6546b64;
+  }
+
+  //----------
+  // tail
+  const uint8_t* tail = (const uint8_t*)(data + (int64_t)nblocks * 4);
+
+  uint32_t k1 = 0;
+
+  switch (len & 3) {
+    case 3:
+      k1 ^= tail[2] << 16;
+    case 2:
+      k1 ^= tail[1] << 8;
+    case 1:
+      k1 ^= tail[0];
+      k1 *= c1;
+      k1 = ROTL32(k1, 15);
+      k1 *= c2;
+      h1 ^= k1;
+  };
+
+  //----------
+  // finalization
+  h1 ^= len;
+
+  h1 = fmix(h1);
+
+  *(uint32_t*)out = h1;
+}
+
+Status MurmurHash3::Compute(OpKernelContext* ctx) const {
+  const Tensor* keys = ctx->Input<Tensor>(0);
+  ONNXRUNTIME_ENFORCE(keys);
+
+  const TensorShape& input_shape = keys->Shape();
+  Tensor* output_tensor = ctx->Output(0, input_shape);
+
+  auto keys_type = keys->DataType();
+  int input_element_bytes = static_cast<int>(keys->DataType()->Size());
+  int output_element_bytes = static_cast<int>(output_tensor->DataType()->Size());
+  auto input_count = input_shape.Size();
+  for (int i = 0; i < input_count; ++i) {
+    if (DataTypeImpl::GetType<std::string>() == keys_type) {
+      auto input = keys->DataRaw();
+      auto output = output_tensor->MutableDataRaw();
+      auto input_string = reinterpret_cast<const std::string*>(input)[i];
+      input_string.length();
```

Comment:
? do you need the output of this?

Reply:
this is a mistake

---
In reply to: [241616367](https://github.com/Microsoft/onnxruntime/pull/174#discussion_r241616367) [](ancestors = 241616367)